cli/diagnostics.rs

```diff
@@ -0,0 +1,640 @@
+// Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
+use crate::ansi;
+use serde_json;
+use serde_json::value::Value;
+use std::fmt;
+
+// A trait which specifies parts of a diagnostic like item needs to be able to
+// generate to conform its display to other diagnostic like items
+pub trait DisplayFormatter {
+  fn format_category_and_code(&self) -> String;
+  fn format_message(&self, level: usize) -> String;
+  fn format_related_info(&self) -> String;
+  fn format_source_line(&self, level: usize) -> String;
+  fn format_source_name(&self, level: usize) -> String;
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct Diagnostic {
+  pub items: Vec<DiagnosticItem>,
+}
+
+impl Diagnostic {
+  /// Take a JSON value and attempt to map it to a
+  pub fn from_json_value(v: &serde_json::Value) -> Option<Self> {
+    if !v.is_object() {
+      return None;
+    }
+    let obj = v.as_object().unwrap();
+
+    let mut items = Vec::<DiagnosticItem>::new();
+    let items_v = &obj["items"];
+    if items_v.is_array() {
+      let items_values = items_v.as_array().unwrap();
+
+      for item_v in items_values {
+        items.push(DiagnosticItem::from_json_value(item_v));
+      }
+    }
+
+    Some(Self { items })
+  }
+
+  pub fn from_emit_result(json_str: &str) -> Option<Self> {
+    let v = serde_json::from_str::<serde_json::Value>(json_str)
+      .expect("Error decoding JSON string.");
+    let diagnostics_o = v.get("diagnostics");
+    if let Some(diagnostics_v) = diagnostics_o {
+      return Self::from_json_value(diagnostics_v);
+    }
+
+    None
+  }
+}
+
+impl fmt::Display for Diagnostic {
+  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+    let mut i = 0;
+    for item in &self.items {
+      if i > 0 {
+        writeln!(f)?;
+      }
+      write!(f, "{}", item.to_string())?;
+      i += 1;
+    }
+
+    if i > 1 {
+      write!(f, "\n\nFound {} errors.\n", i)?;
+    }
+
+    Ok(())
+  }
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct DiagnosticItem {
+  pub message: String,
+  pub message_chain: Option<Box<DiagnosticMessageChain>>,
+  pub related_information: Option<Vec<DiagnosticItem>>,
+  pub source_line: Option<String>,
+  pub line_number: Option<i64>,
+  pub script_resource_name: Option<String>,
+  pub start_position: Option<i64>,
+  pub end_position: Option<i64>,
+  pub category: DiagnosticCategory,
+  pub code: i64,
+  pub start_column: Option<i64>,
+  pub end_column: Option<i64>,
```

Comment:
Can you add some triple slash comments for whether the following are 0-indexed or 1-indexed: line_number, start_position, end_position, start_column, end_column

Add a comment to `code` saying that it's a TypeScript code maybe with a link to diagnosticMessages.json (https://github.com/microsoft/TypeScript/blob/master/src/compiler/diagnosticMessages.json)